### PR TITLE
[WSL-315] Refactoring end-to-end tests

### PR DESCRIPTION
**Pull request by EduardGomezEscandell**
This refactor aims to make the tests more idiomatic.

### Changes

These changes are so interlinked that creating a sequence of commits is a fool's errand. Instead, the refactor is one big commit:
- Changed some filenames
- Removed Tester struct
- Switched to using `require` package
- Created `test_cases.go`with reusable assertion sets that can be run as parallel subtests.
- Created test for `launcher.exe install --root --ui=none`.

The change to the CI can be isolated, I suppose:
- Modified CI to point to new files.

### CODE REVIEW

**Comment by EduardGomezEscandell** [comment]
Here is some work left to do, but they all imply an expansion of functionality beyond the scope of this refactor. They should be solved in a follow-up PR.
- In TestBasicSetupFallbackUI: Try to conrtol stdin/stdout to avoid the need for `--root` installation. https://github.com/ubuntu/WSL/blob/dc8586a4cbfc4aea32faf061e94dca4dec6b3af0/e2e/test-runner/basic_setup_test.go#L43-L44
- In TestDefaultExperience: Find out how to control this remotely, or at least read stdout by redirecting it to a file. Also, analyse why this test fails on Azure but not on a regular machine. https://github.com/ubuntu/WSL/blob/dc8586a4cbfc4aea32faf061e94dca4dec6b3af0/e2e/test-runner/default_experience_test.go#L23-L25

File: e2e/test-runner/test_cases.go [hunk 1]
```diff
@@ -140,3 +176,20 @@ func testUpgradePolicyIdempotent(t *testing.T) {
 
 	require.Equal(t, string(wantsDate), string(gotDate), "Launcher is modifying release upgrade every boot")
 }
+
+// systemdExpected returns true if systemd is expected to be enabled by default on this distro
+func systemdExpected() bool {
```
Comment (didrocks) [inline e2e/test-runner/test_cases.go]: Nitpick: I would even make it a sentence `systemdIsExpected()`

File: e2e/test-runner/default_experience_test.go [hunk 2]
```diff
@@ -17,6 +17,17 @@ func TestDefaultExperience(t *testing.T) {
 	t.Skip("Skipped: fails in Azure") // TODO: Fix
 	wslSetup(t)
 
+	var installCompleted bool
+	t.Cleanup(func() {
+		if installCompleted {
```
Comment (didrocks) [inline e2e/test-runner/default_experience_test.go]: I think that reads better in one line: `if installCompleted && !t.Failed()`.

File: e2e/test-runner/test_utils.go [hunk 3]
```diff
@@ -30,15 +32,8 @@ func wslSetup(t *testing.T) {
...
+	removeOldLogs(t)
 
-	rootDir := os.Getenv(constants.LauncherRepoEnvVar)
-	clientLogFullPath := filepath.Join(rootDir, clientLogPath)
-	err := os.Remove(clientLogFullPath)
-	if err != nil {
-		require.ErrorIs(t, err, os.ErrNotExist, "Failed to remove old install log at %q.", clientLogFullPath)
-	}
-
-	// Attempts to unregister the instance.
```
Comment (didrocks) [inline e2e/test-runner/test_utils.go]: Was the comment removal intended?

File: e2e/test-runner/test_utils.go [hunk 4]
```diff
@@ -49,36 +44,52 @@ func wslSetup(t *testing.T) {
...
-		out, err = wslCommand(ctx, "cat", ServerLogPath).CombinedOutput()
-		t.Logf("%s", out)
-		if err != nil {
-			t.Logf("Failed to retrieve server debug log: %v", err)
-		}
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	out, err := wslCommand(ctx, "cat", subiquityAnswerFile).CombinedOutput()
+	if err != nil {
+		t.Logf("Failed to retrieve subiquity answer file: %v\n%s", err, out)
```
Comment (didrocks) [inline e2e/test-runner/test_utils.go]: Don’t print out if there is an error. The first argument content validity is not ensure in generally in Go when a function returns an error

File: e2e/test-runner/test_utils.go [hunk 5]
```diff
@@ -49,36 +44,52 @@ func wslSetup(t *testing.T) {
...
-			t.Logf("Failed to retrieve client debug log: %v", err)
-		}
-	})
+	ctx, cancel = context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	out, err = wslCommand(ctx, "cat", ServerLogPath).CombinedOutput()
+	if err != nil {
+		t.Logf("Failed to retrieve server debug log: %v\n%s", err, out)
+	} else {
+		s = s + fmt.Sprintf("\n==== Server Debug Log ====\n%s\n", out)
```
Comment (didrocks) [inline e2e/test-runner/test_utils.go]: ditto

File: e2e/test-runner/test_utils.go [hunk 6]
```diff
@@ -49,36 +44,52 @@ func wslSetup(t *testing.T) {
...
+	}
+
+	out, err = ioutil.ReadFile(clientLogFullPath)
+	if err != nil {
+		t.Logf("Failed to retrieve client debug log: %v\n%s", err, out)
+	} else {
+		s = s + fmt.Sprintf("\n==== Client Debug Log ====\n%s\n", out)
```
Comment (didrocks) [inline e2e/test-runner/test_utils.go]: ditto

File: e2e/test-runner/test_cases.go [hunk 7]
```diff
@@ -15,7 +15,7 @@ import (
 )
 
 // testUserNotRoot ensures the default user is not root.
-func testUserNotRoot(t *testing.T) {
+func testUserNotRoot(t *testing.T) { //nolint: thelper
```
Comment (didrocks) [inline e2e/test-runner/test_cases.go]: I would add an additional comment like ("This is a test"). Same below, but I won’t repeat it :)

File: e2e/test-runner/test_cases.go [hunk 8]
```diff
@@ -85,13 +85,17 @@ func testSystemdEnabled(t *testing.T) {
...
+func testSystemdUnits(t *testing.T) { //nolint: thelper
 	if !systemdExpected() {
 		// Enable systemd in config file
-		wslCommandAsUser(context.Background(), "root", "bash", "-c", `printf '\n[boot]\nsystemd=true\n' >> /etc/wsl.conf`)
+		out, err := wslCommandAsUser(context.Background(), "root", "bash", "-c", `printf '\n[boot]\nsystemd=true\n' >> /etc/wsl.conf`).CombinedOutput()
+		require.NoError(t, err, "Failed to enable systemd: %s", out)
 		// Restore config file
-		defer wslCommandAsUser(context.Background(), "root", "bash", "-c", `head -n -2 "/etc/wsl.conf" > "/etc/wsl.conf"`)
+		t.Cleanup(func() {
+			out, err := wslCommandAsUser(context.Background(), "root", "bash", "-c", `head -n -2 "/etc/wsl.conf" > "/etc/wsl.conf"`).CombinedOutput()
```
Comment (didrocks) [inline e2e/test-runner/test_cases.go]: I’m not really confortable with that Cleanup approach. Nothing really ensure that the first two lines are what matters in the future.

Can we rather save the original vanilla file somewhere and restore it?

File: e2e/test-runner/test_utils.go [hunk 9]
```diff
@@ -65,14 +65,14 @@ func subiquityLogs(t *testing.T) string {
...
 	defer cancel()
-	out, err = wslCommand(ctx, "cat", ServerLogPath).CombinedOutput()
+	out, err = wslCommand(ctx, "cat", serverLogPath).CombinedOutput()
 	if err != nil {
 		t.Logf("Failed to retrieve server debug log: %v\n%s", err, out)
 	} else {
 		s = s + fmt.Sprintf("\n==== Server Debug Log ====\n%s\n", out)
 	}
 
-	out, err = ioutil.ReadFile(clientLogFullPath)
```
Comment (didrocks) [inline e2e/test-runner/test_utils.go]: oh sorry, I didn’t spot you were using ioutil during our peer programming :) Be warned, as this change is like a couple of Go release ago, you will find a lot of examples still using ioutil, which is now deprecated (all functionalities have similar helpers in different other packages)

File: e2e/launchertester/test_cases.go [hunk 10]
```diff
@@ -43,7 +43,7 @@ func testCorrectReleaseRootfs(t *testing.T) { //nolint: thelper
 		"Ubuntu-22.04":   "22.04",
 		"Ubuntu-20.04":   "20.04",
 		"Ubuntu-18.04":   "18.04",
-		"Ubuntu-Preview": "23.04",
+		"Ubuntu-Preview": "22.10",
```
Comment (didrocks) [inline e2e/launchertester/test_cases.go]: oh, why?

File: e2e/launchertester/test_cases.go [hunk 11]
```diff
@@ -43,7 +43,7 @@ func testCorrectReleaseRootfs(t *testing.T) { //nolint: thelper
 		"Ubuntu-22.04":   "22.04",
 		"Ubuntu-20.04":   "20.04",
 		"Ubuntu-18.04":   "18.04",
-		"Ubuntu-Preview": "22.10",
+		"Ubuntu-Preview": "23.04",
```
Comment (didrocks) [inline e2e/launchertester/test_cases.go:46]: Ack :p

Comment (EduardGomezEscandell) [reply e2e/launchertester/test_cases.go:46]: We had a pull request race condition 😆

Comment (didrocks) [reply e2e/launchertester/test_cases.go:46]: Exactly :-)

File: e2e/test-runner/basic_setup_test.go [hunk 12]
```diff
@@ -12,7 +12,7 @@ import (
 func TestBasicSetupGUI(t *testing.T) {
 	wslSetup(t)
 
-	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
```
[the remark this answers is not in the document]
Comment (EduardGomezEscandell) [reply e2e/test-runner/basic_setup_test.go]: Makes sense. I also added timeouts to other test cases to be more consistent.

File: e2e/test-runner/test_utils.go (hunk 3 above)
Comment (EduardGomezEscandell) [reply e2e/test-runner/test_utils.go]: Yeah, I felt it was unnecessary given that the cleanup function is short and (from my perspective) quite clear. The comment added no extra info.

File: e2e/test-runner/test_utils.go (hunk 4 above)
Comment (EduardGomezEscandell) [reply e2e/test-runner/test_utils.go]: I know it's bad in most other contexts, but isn't this quite standard for `Cmd.CombinedOutput`? What's the idiomatic way of reading the error message of a failed command?

File: e2e/test-runner/test_cases.go (hunk 8 above)
Comment (EduardGomezEscandell) [reply e2e/test-runner/test_cases.go]: Yeah that makes much more sense. On it.

File: e2e/test-runner/test_utils.go (hunk 4 above)
Comment (didrocks) [reply e2e/test-runner/test_utils.go]: Ah sorry, this is `CombinedOutput()`, so yeah, this is one exception when we don’t redirect stderr before. So, let’s keep it this way.

File: e2e/test-runner/test_utils.go (hunk 9 above)
Comment (EduardGomezEscandell) [reply e2e/test-runner/test_utils.go]: Yep, the linter also complained so I fixed it later in this commit: 096600d8c41e3b697c1cf74f3159ff32b6b3fe5d

File: e2e/test-runner/default_experience_test.go (hunk 2 above)
Comment (didrocks) [reply e2e/test-runner/default_experience_test.go]: oh right, that was || to get into the condition :p

**Review by didrocks — APPROVED** [review]
Looks excellent now! Thanks for the fixes. Really nice work :)

**Pull request merged**